pkg/controller/certificaterequests/ca/ca_test.go

```diff
@@ -391,3 +399,168 @@ func runTest(t *testing.T, test testT) {
+				&x509.Certificate{
+					SerialNumber: big.NewInt(1234),
+					IsCA:         true,
+				},
+			))),
+			givenNamespace: "default",
+			assertSignedCert: func(t *testing.T, got *x509.Certificate) {
+				assert.Equal(t, true, got.IsCA)
+			},
+		},
+		"when the Issuer has ocspServers set, it should appear on the signed ca": {
+			givenCAIssuer: gen.Issuer("issuer-1", gen.SetIssuerCA(cmapi.CAIssuer{
+				SecretName:  "secret-1",
+				OCSPServers: []string{"http://ocsp-v3.example.org"},
+			})),
+			givenCR: gen.CertificateRequest("cr-1",
+				gen.SetCertificateRequestCSR(rsaCSR),
+				gen.SetCertificateRequestIssuer(cmmeta.ObjectReference{
+					Name:  "issuer-1",
+					Group: certmanager.GroupName,
+					Kind:  "Issuer",
+				}),
+			),
+			givenCASecret: gen.SecretFrom(gen.Secret("secret-1"), gen.SetSecretNamespace("default"), gen.SetSecretData(secretDataFor(t, rsaPair,
+				&x509.Certificate{
+					SerialNumber: big.NewInt(1234),
+					IsCA:         true,
+				},
+			))),
+			givenNamespace: "default",
+			assertSignedCert: func(t *testing.T, got *x509.Certificate) {
+				assert.Equal(t, []string{"http://ocsp-v3.example.org"}, got.OCSPServer)
+			},
+		},
+		"when the Issuer has crlDistributionPoints set, it should appear on the signed ca ": {
+			givenCAIssuer: gen.Issuer("issuer-1", gen.SetIssuerCA(cmapi.CAIssuer{
+				SecretName:            "secret-1",
+				CRLDistributionPoints: []string{"http://www.example.com/crl/test.crl"},
+			})),
+			givenCR: gen.CertificateRequest("cr-1",
+				gen.SetCertificateRequestIsCA(true),
+				gen.SetCertificateRequestCSR(rsaCSR),
+				gen.SetCertificateRequestIssuer(cmmeta.ObjectReference{
+					Name:  "issuer-1",
+					Group: certmanager.GroupName,
+					Kind:  "Issuer",
+				}),
+			),
+			givenCASecret: gen.SecretFrom(gen.Secret("secret-1"), gen.SetSecretNamespace("default"), gen.SetSecretData(secretDataFor(t, rsaPair,
+				&x509.Certificate{
+					SerialNumber: big.NewInt(1234),
+					IsCA:         true,
+				},
+			))),
+			givenNamespace: "default",
+			assertSignedCert: func(t *testing.T, gotCA *x509.Certificate) {
+				assert.Equal(t, []string{"http://www.example.com/crl/test.crl"}, gotCA.CRLDistributionPoints)
+			},
+		},
+	}
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			rec := &controllertest.FakeRecorder{}
+
+			c := &CA{
+				issuerOptions: controller.IssuerOptions{
+					ClusterResourceNamespace:        "",
+					ClusterIssuerAmbientCredentials: false,
+					IssuerAmbientCredentials:        false,
+				},
+				reporter: util.NewReporter(fixedClock, rec),
+				secretsLister: listers.FakeSecretListerFrom(listers.NewFakeSecretLister(),
+					listers.SetFakeSecretNamespaceListerGet(test.givenCASecret, nil),
+				),
+				templateGenerator: pki.GenerateTemplateFromCertificateRequest,
+			}
+
+			gotIssueResp, gotErr := c.Sign(context.Background(), test.givenCR, test.givenCAIssuer)
+			if test.wantErr != "" {
+				assert.EqualError(t, gotErr, test.wantErr)
+				return
```

Comment:
Suggestion: is there maybe a way how we can not return here and still run the other checks even when we expect an error (i.e by testing that `gotCert` is nil etc)? 
I feel like it is a little confusing that we return in a test- but happy to be convinced otherwise.

Reply:
I also realize that returning here potentially "skips" below assertions that might have to be checked even if an error is returned...

Would it make sense to do instead

```go
if test.wantErr != "" {
	assert.EqualError(t, gotErr, test.wantErr)
} else {
	require.NoError(t, gotErr)
	require.NotNil(t, gotIssueResp)
	gotCert, err := pki.DecodeX509CertificateBytes(gotIssueResp.Certificate)
	require.NoError(t, err)
	test.assertSignedCert(t, gotCert)                
}
```

Reply:
I think this is clearer! I am always torn between whether I actually want to check error strings and whether I want to verify for example that `gotIssuerResp` is nil in case of an error.
I personally have a slight preference for not having else blocks and not comparing error strings only types if possible:


```
// given that `wantsErr` is boolean and  `gotErr` is error
if test.wantsErr != (test.gotErr != nil) {
	// unexpected presence error or the lack of it, fail
}
require.Equal(test.IssuerResp, gotIssuerResp) // check both that it's correct when there's no error and nil when there is error 
if gotIssuerResp != nil {
	gotCert, err := pki.DecodeX509CertificateBytes(gotIssueResp.Certificate)
	require.NoError(t, err)
	test.assertSignedCert(t, gotCert)
}          
```

But I don't think this ^ is better, just different, so your fix looks good to me.